.gitignore

```diff
@@ -1,12 +1,10 @@
 target/
 bin/
 semantic.cache
-tmp/*
-!tmp/.gitkeep

```

Comment:
> But I don't see any code changes here that would ensure the existence of tmp/.

We don’t need any such code.

For newly seeded repositories, the `src/ontology/tmp` directory will be automatically created as part of the normal seeding process because we now have a `src/ontology/tmp/.gitkeep` file in the template.

For existing repositories, they will get the `src/ontology/tmp` when their users run the `update_repo` command. And yes, I hear what you said in your PR about the fact that having to run a command to upgrade is bad UX, but that’s the way the ODK has been working all along – it’s always necessary to run `update_repo` after a major upgrade. It may be bad UX but then the entire ODK is guilty of having bad UX. :(

Reply:
>  the `src/ontology/tmp` directory will be automatically created as part of the normal seeding process
Ah, I see. I was thinking that might be the case.

> I hear what you said in your PR about the fact that having to run a command to upgrade is bad UX
Really appreciate you reading through my comments! I'm just happy to communicate that issue even if it's not solved at this time. An idea is popping into my head right now about some means of accounting for version-ups; like perhaps the ODK version at instantiation time is cached, and then when `run.sh` runs, it detects if it's running on a newer version, and then it executes some subset of `update_repo`, like an `rsync`, or maybe just warns the user about needing to run it; that would probably suffice.

Anyway this is good to merge then!

Reply:
> the ODK version at instantiation time is cached, and then when run.sh runs, it detects if it's running on a newer version, and then it executes some subset of update_repo

Automatically updating (even partially) a repository is a no-go. The decision to update should always be left at the discretion of the users.

Any update brings the risk of breaking someone’s workflows (especially _custom_ workflows, on which we have no control), so users should always be in control of when they want to run the risk of updating (for example, if they plan to make a release of their ontology soon, they may want to postpone updating their repo for a few days until after the release).

> maybe just warns the user about needing to run it; that would probably suffice.

Yes.